video_xblock/mixins.py

```diff
@@ -165,86 +195,98 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
-            error_message (dict): Description of error.
-        """
-        log.debug("Getting 3PlayMedia transcripts...")
-        domain = 'https://static.3playmedia.com/'
-        transcripts_3playmedia = requests.get(
-            '{domain}files/{file_id}/transcripts?apikey={api_key}'.format(
-                domain=domain, file_id=file_id, api_key=apikey
-            )
-        ).json()
-
-        errors = isinstance(transcripts_3playmedia, dict) and transcripts_3playmedia.get('errors')
-        if errors:
-            return 'error', {'error_message': u'\n'.join(errors.values())}
-
-        translations = []
-        for transcript in transcripts_3playmedia:
-            tid = transcript.get('id', '')
-            lang_id = transcript.get('language_id')
-            formatted_content = requests.get(
-                '{domain}files/{file_id}/transcripts/{tid}?apikey={api_key}&format_id={format_id}'.format(
-                    domain=domain, file_id=file_id, api_key=apikey, tid=tid,
-                    format_id=TPMApiTranscriptFormatID.WEBVTT
-                )
-            ).text
-            lang_data = TPMApiLanguage(lang_id)
-            lang_label = lang_data.name
-            video_id = self.get_player().media_id(self.href)
-            source = TranscriptSource.THREE_PLAY_MEDIA
+        :return: (generator of OrderedDicts) all transcript's data
+        """
+        feedback, transcripts_list = self.get_3pm_transcripts_list(
+            self.threeplaymedia_file_id, self.threeplaymedia_apikey
+        )
+        log.debug("Fetched 3PM transcripts list results:\n{}".format(feedback))
 
-            reference_name = create_reference_name(lang_label, video_id, source)
+        if feedback['status'] is Status.error:
+            log.error("3PlayMedia transcripts fetching API request has failed!\n{}".format(feedback['message']))
+            raise StopIteration
 
-            _file_name, external_url = self.create_transcript_file(
-                trans_str=formatted_content,
-                reference_name=reference_name
-            )
-            translations.append({
-                "lang": lang_data.iso_639_1_code,
-                "label": lang_label,
-                "url": external_url,
-                "source": source,
-            })
-        log.debug("Got 3PlayMedia transcripts: {}".format(translations))
-        return 'success', translations
+        for transcript_data in transcripts_list:
+            transcript = self.fetch_single_3pm_translation(transcript_data)
+            if transcript is None:
+                raise StopIteration
+            transcript_ordered_dict = transcript._asdict()
+            transcript_ordered_dict['content'] = ''  # we don't want to parse it to JSON
+            yield transcript_ordered_dict
 
-    @XBlock.json_handler
-    def get_transcripts_3playmedia_api_handler(self, data, _suffix=''):
+    def get_3pm_transcripts_list(self, file_id, apikey):
```

Comment:
Here we fetch up to date list of available translations for a given File ID and serve it back to the Video player.